### PR TITLE
[WIP] Basic multiplayer architecture

### PR DESCRIPTION
**Pull request by thomasgaozx**
- Packet I/O
- Delta encoding structures
- Snapshot history structure

### CODE REVIEW

File: src/net/types.h [hunk 1]
```diff
@@ -28,6 +28,30 @@ struct NetworkOptions {
     int port { 0 };
 };
 
+struct BaseStatus {
+    
+};
+
+// dummy structure for testing
+struct CreatureStatus : public BaseStatus {
+    CreatureStatus() : x(0), y(0), z(0), heading(0), // dummy
```
Comment (seedhartha) [inline src/net/types.h:37]: This can be compacted into:

```cpp
struct CreatureStatus : public BaseStatus {
    CreatureStatus() = default; // redundant

    float x { 0.0f };
    double heading { 0.0 };
    uint8_t anim { 0 };
    uint16_t equipment1 { 0 };
};
```

File: src/net/types.h [hunk 2]
```diff
@@ -28,6 +28,30 @@ struct NetworkOptions {
     int port { 0 };
 };
 
+struct BaseStatus {
+    
+};
+
+// dummy structure for testing
+struct CreatureStatus : public BaseStatus {
```
Comment (seedhartha) [inline src/net/types.h:36]: Can you put these structs into a separate header file? I now lean towards "types.h" files being containers for constants and enums.

**Comment by seedhartha** [comment]
Abundant use of templates here concerns me. Templates have their uses - particularly, when implementing generic containers and algorithms - but they hinder readability and compilation time. How much duplication will there be if templates were replaced with static types?

File: src/net/packet.h [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+namespace reone {
+
+namespace net {
+
+bool isBigEndian();
+void unitTestSetEndian(bool isBig);
+
+template <typename T>
+std::enable_if_t< std::is_integral_v<T> || std::is_floating_point_v<T>> swapByte(T &val) {
```
Comment (seedhartha) [inline src/net/packet.h:32]: Single responsibility principle. "common/endianutil" should be extended as needed and used here instead.

File: src/net/objects.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+#include <cstdint>
+
+namespace reone {
+
+namespace net {
+
+struct BaseStatus { };
+
+// dummy structure for testing
+struct CreatureStatus : public BaseStatus {
```
Comment (seedhartha) [inline src/net/objects.h:29]: libnet was intended to be lower level than that. Notion of creature state (especially equipment, faction and etc.) is game logic and belongs to libmp.

File: src/net/objects.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    CreatureStatus() = default;
```
Comment (seedhartha) [inline src/net/objects.h:30]: This constructor is redundant, I think. Also, member fields are now uninitialized. This syntax should be used here:

```cpp
float x { 0.0f };
```

File: src/common/endianutil.cpp [hunk 6]
```diff
@@ -17,77 +17,13 @@
...
-}
-
-template <>
-void swapBytes(float &val) {
-    uint32_t uintVal;
-    memcpy(&uintVal, &val, 4);
-    swapBytes(uintVal);
-    memcpy(&val, &uintVal, 4);
-}
+uint16_t test = 0x1100;
```
Comment (seedhartha) [inline src/common/endianutil.cpp:22]: That's clever 👍 As per code style, global variables should bein with "g_". Also, how about a separate init function?

```cpp
Endianess initEndianess() {
    uint16_t test = 0x1100;
    return static_cast<bool>(*(char*)&test) ? Endianess::Big : Endianess::Little;
}

const Endianess g_end = initEndianess();
```

File: src/common/endianutil.h [hunk 7]
```diff
@@ -17,9 +17,24 @@
...
+/* only permits fundamental integral types or extended integral types */
+template <typename T>
+std::enable_if_t<std::is_integral<T>::value || std::is_floating_point<T>::value> swapBytes(T &val) {
+	char *c = reinterpret_cast<char*>(&val);
+	for (size_t i=0; i*2 < sizeof(T); ++i) std::swap(c[i], c[sizeof(T) - i - 1]);
+}
+
 template <class T>
-void swapBytes(T &val);
+void fixEndianess(T& val,  bool isBigEndian) {
```
Comment (seedhartha) [inline src/common/endianutil.h:36]: I don't think it belongs here. The whole point of `fixEndianess` was "swap bytes if endianess of the STREAM differs from endianess of the system".

File: src/net/objects.h (hunk 4 above)
Comment (thomasgaozx) [reply src/net/objects.h:29]: I'll move `snapshot.h` and `snapshot.cpp` to libmp as well.

File: src/common/endianutil.h (hunk 7 above)
Comment (thomasgaozx) [reply src/common/endianutil.h:36]: Interesting, I think I misunderstood its point completely ...

Comment (thomasgaozx) [reply src/common/endianutil.h:36]: What about this?

```c++
template <class T>
void fixEndianess(T& val,  Endianess to);
```

Comment (seedhartha) [reply src/common/endianutil.h:36]: Given we're on x86 and want to read a native (LE) int from a BE stream, "to" would not be a correct name in this example. Also, "fix" is an ambiguous term. Maybe `swapBytesIfNotSystemEndianess` will do?

File: src/common/endianutil.cpp [hunk 8]
```diff
@@ -17,77 +17,21 @@
...
-    val = ((val >> 56) & 0xff) |
-        (((val >> 48) & 0xff) << 8) |
-        (((val >> 40) & 0xff) << 16) |
-        (((val >> 32) & 0xff) << 24) |
-        (((val >> 24) & 0xff) << 32) |
-        (((val >> 16) & 0xff) << 40) |
-        (((val >> 8) & 0xff) << 48) |
-        ((val & 0xff) << 56);
-}
+Endianess _end = initEndianess();
```
Comment (seedhartha) [inline src/common/endianutil.cpp:27]: `Endianess g_end = initEndianess()`, this is a global variable. Might be a good idea to make it static too (to put it in file, not global scope).

File: src/common/endianutil.cpp [hunk 9]
```diff
@@ -17,77 +17,21 @@
...
+    return _end;
 }
 
-template <>
-void swapBytes(double &val) {
-    uint64_t uintVal;
-    memcpy(&uintVal, &val, 8);
-    swapBytes(uintVal);
-    memcpy(&val, &uintVal, 8);
+void unitTestSetEndianess(Endianess en) {
```
Comment (seedhartha) [inline src/common/endianutil.cpp:33]: nit: Generally, using mocks is preferred to including testing handles in production code (see [FakeIt](https://github.com/eranpeer/FakeIt) library). This is better than dragging in another dependency though.

File: src/common/streamwriter.cpp [hunk 10]
```diff
@@ -26,6 +26,14 @@ using namespace std;
 
 namespace reone {
 
+template <class T>
+void put(T val, std::shared_ptr<std::ostream> &stream, Endianess en) {
```
Comment (seedhartha) [inline src/common/streamwriter.cpp:30]: `static` to put it in file, not global scope.

File: src/common/endianutil.cpp (hunk 8 above)
Comment (thomasgaozx) [reply src/common/endianutil.cpp:27]: What about static inside function:

```c++
Endianess getEndianess() {
    static Endianess _end = initEndianess();
    return _end;
}
```

File: src/common/streamwriter.h [hunk 11]
```diff
@@ -34,20 +35,30 @@ class StreamWriter {
...
+        _stream->write(reinterpret_cast<char*>(&obj), sizeof(T));
+    }
+
+    template<typename T>
+    StreamWriter &operator<<(T obj) {
+        write(obj);
+        return *this;
+    }
+
+    const std::shared_ptr<std::ostream> &getStream();
```
Comment (seedhartha) [inline src/common/streamwriter.h:54]: Although slightly more efficient, returning a reference here breaks encapsulation, because now callers can change _stream directly. Should be: `std::shared_ptr<std::ostream> stream() const`

Comment (thomasgaozx) [reply src/common/streamwriter.h:54]: Yeah I believe so

File: src/common/endianutil.cpp (hunk 8 above)
Comment (seedhartha) [reply src/common/endianutil.cpp:27]: Fine by me, but then it's just `end`. Or `endianess` to remove ambiguity.

File: src/mp/snapshot.h [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "objects.h"
+#include "../common/streamreader.h"
+#include "../common/streamwriter.h"
+
+namespace reone {
+
+namespace net {
+
+constexpr uint32_t g_ServerMaxLen = 64;
```
Comment (seedhartha) [inline src/mp/snapshot.h:35]: These are constants, not global variables: `kServerMaxLen`, `kClientMaxLen`.

File: src/mp/snapshot.h [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    uint32_t seq;
+    uint32_t tick;
+};
+
+/* 
+* memory-efficient (cough cough) snapshot system
+* sort of like a skip list */
+class History {
+public:
+    History(uint32_t maxLen) : _snapshots{}, _top{}, _seq{ 0 }, _maxLen(maxLen), _activeObjects{} {
```
Comment (seedhartha) [inline src/mp/snapshot.h:57]: No need to explicitly initialize STL variables, i.e. `_snapshots` and `_activeObjects`. Primitive variables I would initialize in-place, e.g. `uint32_t _seq { 0 }`.

File: src/mp/snapshot.cpp [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+#include "../net/delta.h"
+
+using namespace std;
+
+namespace reone {
+
+namespace net {
+
+void History::insert(uint32_t id) {
+    auto cptr = std::make_unique<CreatureStatus>();
```
Comment (seedhartha) [inline src/mp/snapshot.cpp:29]: `std::` is redundant here and below.

File: src/common/endianutil.h [hunk 15]
```diff
@@ -17,9 +17,24 @@
...
+/* only permits fundamental integral types or extended integral types */
+template <typename T>
+std::enable_if_t<std::is_integral<T>::value || std::is_floating_point<T>::value> swapBytes(T &val) {
+	char *c = reinterpret_cast<char*>(&val);
+	for (size_t i=0; i*2 < sizeof(T); ++i) std::swap(c[i], c[sizeof(T) - i - 1]);
+}
+
 template <class T>
-void swapBytes(T &val);
+void swapBytesIfNotSystemEndianess(T& val, Endianess to) {
```
Comment (seedhartha) [inline src/common/endianutil.h:36]: `to` is not always correct, `other` is more fitting.

File: src/common/streamwriter.cpp [hunk 16]
```diff
@@ -26,6 +26,14 @@ using namespace std;
 
 namespace reone {
 
+template <class T>
+static void put(T val, std::shared_ptr<std::ostream> &stream, Endianess en) {
```
Comment (seedhartha) [inline src/common/streamwriter.cpp:30]: `std::` is redundant.

File: src/mp/snapshot.h [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+namespace reone {
+
+namespace net {
+
+constexpr uint32_t kServerMaxLen = 64;
+constexpr uint32_t kClientMaxLen = 3;  // may increase if too tight
+
+struct Snapshot {
+    Snapshot(uint32_t seq, uint32_t tick) : seq{ seq }, tick{ tick } {};
```
Comment (seedhartha) [inline src/mp/snapshot.h:39]: Let's use this syntax when variables are constructor initialized: `seq(seq), tick(tick)`

File: src/common/streamreader.cpp [hunk 18]
```diff
@@ -135,6 +124,14 @@ string StreamReader::getString(int len) {
     return move(val);
 }
 
+void StreamReader::read(char &obj) {
+    _stream->get(obj);
+}
+
+void StreamReader::read(std::string &obj) {
```
Comment (seedhartha) [inline src/common/streamreader.cpp:131]: `std::` is redundant.

**Comment by seedhartha** [comment]
There are multiple formatting issues with this PR. Please check for the following:

- When defining functions, opening brace should NOT be on a separate line.
- When declaring references (and pointers), add whitespace between a type name and an ampersand, e.g. `type &varName`.
- Let's use the following syntax for declaring template parameters: `template <class T>`.

File: src/net/delta.cpp [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+    readDeltaBuff(getSerPtr(obj), reader);
+}
+
+// https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl
+#define DECL_SERIAL(T, ...) auto getSerPtr(T &obj) \
+    { return make_tuple(__VA_ARGS__); } \
+    template void writeDelta<T>(StreamWriter &writer, T &cur, T &old); \
+    template void readDelta<T>(StreamReader &reader, T &obj);
+
+DECL_SERIAL(CreatureStatus, &obj.x, &obj.y, &obj.z, &obj.heading, &obj.anim, &obj.animframe, &obj.faction);
```
Comment (seedhartha) [inline src/net/delta.cpp:118]: libnet is lower level than libmp, so it shouldn't have to depend on `CreatureStatus`.

File: src/net/delta.cpp [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+    readDeltaImpl(reader, move(obj), header, make_index_sequence<sizeof...(Args)>());
+}
+
+template <typename T>
+void readDelta(StreamReader &reader, T &obj) {
+    readDeltaBuff(getSerPtr(obj), reader);
+}
+
+// https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl
+#define DECL_SERIAL(T, ...) auto getSerPtr(T &obj) \
```
Comment (seedhartha) [inline src/net/delta.cpp:113]: `getSerPtr` is rather ambiguous. How about `getSerial(Member)Pointers`?

File: src/net/delta.cpp [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+
+template <size_t I, typename ... Args>
+void diffHeaderIter(const tuple<Args...> &left, const tuple<Args...> &right, uint16_t &header) {
+    if (get<I>(left) != get<I>(right)) {
+        header |= 1 << I;
+    }
+}
+
+template <size_t ... Is, typename ... Args>
+uint16_t diffHeader(const tuple<Args...> &left, const tuple<Args...> &right, integer_sequence<size_t, Is...>) {
```
Comment (seedhartha) [inline src/net/delta.cpp:54]: Can we give `tuple<Args...>` a type name, e.g. `typedef tuple<T...> SerialMemberPointers<T>`?

File: src/net/delta.cpp [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+template <typename T>
+void writeDelta(StreamWriter &writer, T &cur, T &old) {
+    writeDeltaBuff(getSerial(cur), getSerial(old), writer);
+}
+
+/* r-value to discourage outside scope usage*/
+template <typename ... Args>
+void readDeltaBuff(tuple<Args*...> &&obj, StreamReader &reader) {
```
Comment (seedhartha) [inline src/net/delta.cpp:101]: `Buff` here is not self-descriptive. How about `readDeltaWithHeader`?

File: src/mp/delta.cpp [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+auto getSerImpl(TuplePtr &&seqPtr, integer_sequence<size_t, Is...>) {
+    return make_tuple(*(get<Is>(seqPtr))...);
+}
+
+template <class T>
+auto getSerial(T &obj) {
+    return getSerImpl(getSerialPointers(obj), make_index_sequence<tuple_size<decltype(getSerialPointers(obj))>::value>());
+}
+
+/* helper functions */
```
Comment (seedhartha) [inline src/mp/delta.cpp:42]: Let's use C++-style comments for code folding and implementation notes, and Javadoc-style comments for documentation. See [https://github.com/seedhartha/reone/wiki/Code-Style#comments](https://github.com/seedhartha/reone/wiki/Code-Style#comments).

**Comment by seedhartha** [comment]
If we're keeping the templates, then delta compression code needs solid documentation comments, including the "why" and "how", because right now it is incredibly hard to understand.

File: src/common/streamreader.h [hunk 24]
```diff
@@ -47,6 +48,21 @@ class StreamReader {
     std::string getCString();
     std::string getString(int len);
 
+    void read(char &obj); // alias for getByte
```
Comment (seedhartha) [inline src/common/streamreader.h:51]: What's the point of so many different "read" interfaces? I'd much rather keep the original functions and add a generic `T get()` function for delta compression. Not a fan of the `>>` operator in particular.

File: src/common/streamwriter.h [hunk 25]
```diff
@@ -34,20 +35,30 @@ class StreamWriter {
     void putInt64(int64_t val);
     void putCString(const std::string &str);
 
+    void write(char obj); // alias for putByte
```
Comment (seedhartha) [inline src/common/streamwriter.h:38]: Same: `put`, `write` and `<<` do essentially the same thing. I don't see why is that necessary.

File: src/common/streamreader.h (hunk 24 above)
Comment (thomasgaozx) [reply src/common/streamreader.h:51]: My plan was to add functions I need for this PR without replacing old functions to minimize conflicts. And potentially update/remove old functions in a separate PR.

```c++
    template <class T>
    void read(T &obj) {
        _stream->read(reinterpret_cast<char*>(&obj), sizeof(T));
        swapBytesIfNotSystemEndianess(obj, _endianess);
    }
```

is necessary here for all integral types (so I don't have to explicitly instantiate all of them in .cpp file).
While `void read(char &obj);` may be redundant here, it bypasses byte-swapping loop from the generic function.

What's wrong with `>>` operator?

**Comment by seedhartha** [comment]
Closing this PR because it is getting stale and is likely to be rewritten.
